flow/table_test.go

```diff
@@ -0,0 +1,300 @@
+
+func TestFlowTable_expire(t *testing.T) {
+	const MaxInt64 = int64(^uint64(0) >> 1)
+	ft := NewFlowTableComplex(t)
+
+	fc := MyTestFlowCounter{}
+	beforeNbFlow := fc.NbFlow
+	ft.expire(fc.expireCallback, 0)
+	afterNbFlow := fc.NbFlow
+	if beforeNbFlow != 0 && afterNbFlow != 0 {
+		t.Error("we should not expire a flow")
+	}
+
+	fc = MyTestFlowCounter{}
+	beforeNbFlow = fc.NbFlow
+	ft.expire(fc.expireCallback, MaxInt64)
+	afterNbFlow = fc.NbFlow
+	if beforeNbFlow != 0 && afterNbFlow != 10 {
+		t.Error("we should expire all flows")
+	}
+}
+
+func TestFlowTable_AsyncExpire(t *testing.T) {
+	t.Skip()
+}
+
+func TestFlowTable_IsExist(t *testing.T) {
+	ft := NewFlowTableSimple(t)
+	flow := &Flow{}
+	flow.UUID = "1234"
+	if ft.IsExist(flow) == false {
+		t.Fail()
+	}
+	flow.UUID = "12345"
+	if ft.IsExist(flow) == true {
+		t.Fail()
+	}
+}
+
+func TestFlowTable_GetFlow(t *testing.T) {
+	ft := NewFlowTableComplex(t)
+	flows := generateFlows(t, ft)
+	if len(flows) != 10 {
+		t.Error("missing some flows ", len(flows))
+	}
+	forgeEthIPTCP(t, int64(1234))
+	_, new := ft.GetFlow("abcd")
+	if !new {
+		t.Error("Collision in the FlowTable, should be new")
+	}
+	forgeEthIPTCP(t, int64(1234))
+	_, new = ft.GetFlow("abcd")
+	if new {
+		t.Error("Collision in the FlowTable, should be an update")
+	}
+	forgeEthIPTCP(t, int64(1234))
+	_, new = ft.GetFlow("abcde")
+	if !new {
+		t.Error("Collision in the FlowTable, should be a new flow")
+	}
+}
+
+func isJSON(s string) bool {
+	var js map[string]interface{}
+	return json.Unmarshal([]byte(s), &js) == nil
+}
+
+func TestFlowTable_JSONFlowConversationEthernetPath(t *testing.T) {
+	ft := NewFlowTableComplex(t)
+	statStr := ft.JSONFlowConversationEthernetPath()
+	if statStr == `{"nodes":[],"links":[]}` {
+		t.Error("stat should not be empty")
+	}
+	if !isJSON(statStr) {
+		t.Error("stat should be JSON format")
+	}
+
+	decoded := new(interface{})
+	if err := json.Unmarshal([]byte(statStr), decoded); err != nil {
+		t.Error("JSON parsing failed:", err)
+	}
+
+	schema := &j.Object{Properties: []j.ObjectItem{
+		{"nodes", &j.Array{Each: &j.Object{Properties: []j.ObjectItem{
+			{"name", &j.String{MinLen: 1}},
+			{"group", &j.Number{Min: 0, Max: 20}}}}},
+		},
+		{"links", &j.Array{Each: &j.Object{Properties: []j.ObjectItem{
+			{"source", &j.Number{Min: 0, Max: 20}},
+			{"target", &j.Number{Min: 0, Max: 20}},
+			{"value", &j.Number{Min: 0, Max: 9999}}}}},
+		},
+	}}
+	if path, err := schema.Validate(decoded); err != nil {
+		t.Errorf("Failed (%s). Path: %s", err, path)
+	}
+}
+
+func TestFlowTable_NewFlowTableFromFlows(t *testing.T) {
+	ft := NewFlowTableComplex(t)
+	var flows []*Flow
+	for _, f := range ft.table {
+		flow := *f
+		flows = append(flows, &flow)
+	}
+	ft2 := ft.NewFlowTableFromFlows(flows)
+	if len(ft.table) != len(ft2.table) {
+		t.Error("NewFlowTable(copy) are not the same size")
+	}
+	flows = flows[:0]
+	for _, f := range ft.table {
+		flows = append(flows, f)
+	}
+	ft3 := ft.NewFlowTableFromFlows(flows)
```

Comment:
Maybe I missed something, I don't understand the purpose of this test. It seems to me that your testing the same thing than L271 except that flow are not copied before ? so the NewFlowTableFromFlows that you are testing process always a pointer ?